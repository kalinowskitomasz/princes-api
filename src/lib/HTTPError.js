export default class HTTPError extends Error {
    constructor(message, status) {
        super(message)
        this.status = status
    }
}
