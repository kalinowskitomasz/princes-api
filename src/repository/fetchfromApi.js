const numberOfRetries = 3

async function fetchRetry(url, options = {}, retries) {
    let res
    try {
        res = await fetch(url, options)
    } catch (err) {
        if (retries > 0) {
            console.error(`Request failed: ${url}. Retry.`)
            console.error(err)
            return fetchRetry(url, options, retries - 1)
        }
        throw new Error(err)
    }
    if (res.ok) {
        return res.json()
    }
    if (retries > 0) {
        console.error(`Request failed: ${url} returned ${res.status}. Retry.`)
        return fetchRetry(url, options, retries - 1)
    }
    throw new Error(res.status)
}

export default async function fetchFromApi(url) {
    return fetchRetry(
        url,
        {
            method: 'GET',
            headers: { 'x-api-key': API_KEY, Accept: 'application/json' },
        },
        numberOfRetries
    )
}
