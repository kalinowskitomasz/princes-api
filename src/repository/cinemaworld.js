import fetchFromApi from './fetchFromApi'
import { logError } from '../lib/logger'

const cinemaworld = {
    movies: `${CINEMAWORLD_API}/movies`,
    movie: `${CINEMAWORLD_API}/movie`,
}

export async function getCinemaworldAllMovies() {
    try {
        const response = await fetchFromApi(new URL(cinemaworld.movies))
        return { error: null, data: response.Movies || [] }
    } catch (err) {
        logError('failed to fetch cinemaworld movies', err)
        return { error: err, data: [] }
    }
}

export async function getCinemaworldMovie(id) {
    try {
        const response = await fetchFromApi(
            new URL(`${cinemaworld.movie}/${id}`)
        )
        return { error: null, data: response }
    } catch (err) {
        logError(`failed to fetch cinemaworld movie ${id}`, err)
        return { error: err, data: {} }
    }
}
