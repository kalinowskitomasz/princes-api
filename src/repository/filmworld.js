import fetchFromApi from './fetchFromApi'
import { logError } from '../lib/logger'

const filmworld = {
    movies: `${FILMWORLD_API}/movies`,
    movie: `${FILMWORLD_API}/movie`,
}

export async function getFilmworldAllMovies() {
    try {
        const response = await fetchFromApi(new URL(filmworld.movies))
        return { error: null, data: response.Movies || [] }
    } catch (err) {
        logError('failed to fetch filmworld movies', err)
        return { error: err, data: [] }
    }
}

export async function getFilmworldMovie(id) {
    try {
        const response = await fetchFromApi(new URL(`${filmworld.movie}/${id}`))
        response.Source = 'Film World'
        return { error: null, data: response }
    } catch (err) {
        logError(`failed to fetch filmworld movie ${id}`, err)
        return { error: err, data: {} }
    }
}
