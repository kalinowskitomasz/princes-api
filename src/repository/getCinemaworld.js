import fetchFromApi from './fetchFromApi'
import { logError } from '../lib/logger'

const cinemaworld = {
    movies: new URL(`${CINEMAWORLD_API}/movies`),
}

export async function getCinemaworldMovies() {
    try {
        const response = await fetchFromApi(cinemaworld.movies)
        return { error: null, data: response.Movies || [] }
    } catch (err) {
        logError('failed to fetch filmworld movies')
        return { error: err, data: [] }
    }
}
