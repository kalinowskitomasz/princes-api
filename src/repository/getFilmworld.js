import fetchFromApi from './fetchFromApi'
import { logError } from '../lib/logger'

const filmworld = {
    movies: new URL(`${FILMWORLD_API}/movies`),
}

export async function getFilmworldMovies() {
    try {
        const response = await fetchFromApi(filmworld.movies)
        return { error: null, data: response.Movies || [] }
    } catch (err) {
        logError('failed to fetch filmworld movies')
        return { error: err, data: [] }
    }
}
