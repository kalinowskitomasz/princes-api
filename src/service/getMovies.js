import { StatusCodes } from 'http-status-codes'

import moviesHandler from '../controller/moviesHandler'
import HTTPError from '../lib/HTTPError'
import { getCinemaworldMovies } from '../repository/getCinemaworld'
import { getFilmworldMovies } from '../repository/getFilmworld'

function mergeDataSources(allMovies) {
    const moviesMap = new Map()
    allMovies.forEach(source => {
        source.data.forEach(movie => {
            // Take advantage of the fact that both services use
            // IMDB id with modified prefix
            // revert it back to original for consistency and data integrity
            const imdbId = `tt${movie.ID.slice(2)}`
            if (!moviesMap.has(imdbId)) {
                moviesMap.set(imdbId, { ...movie, ID: imdbId })
            }
        })
    })
    return [...moviesMap.values()]
}

export default async function getMovies() {
    const allMovies = await Promise.all([
        getCinemaworldMovies(),
        getFilmworldMovies(),
    ])

    // if at least one of the sources returns
    // response still could be sent
    let success = false
    allMovies.forEach(response => {
        success = success || !response.error
    })

    if (!success)
        throw new HTTPError('getMovies failed', StatusCodes.BAD_GATEWAY)

    const moviesData = mergeDataSources(allMovies)

    if (moviesData.length === 0)
        throw new HTTPError('no movies found', StatusCodes.NOT_FOUND)

    return moviesData
}
