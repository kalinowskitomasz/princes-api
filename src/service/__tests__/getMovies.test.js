import getMovies from '../getMovies'
import { getCinemaworldMovies } from '../../repository/getCinemaworld'
import { getFilmworldMovies } from '../../repository/getFilmworld'
import cinemaworldResponse from './cinemaworldResponse.json'
import filmworldResponse from './filmworldResponse.json'
import HTTPError from '../../lib/HTTPError'

jest.mock('../../repository/getCinemaworld')
jest.mock('../../repository/getFilmworld')

describe('getMovies', () => {
    beforeEach(() => {
        getCinemaworldMovies.mockReset()
        getFilmworldMovies.mockReset()
    })

    test('correct responses from both sources', async () => {
        getCinemaworldMovies.mockResolvedValue(cinemaworldResponse)
        getFilmworldMovies.mockResolvedValue(filmworldResponse)

        await expect(getMovies()).resolves.toMatchSnapshot()
    })

    test('correct responses from both only once source', async () => {
        getCinemaworldMovies.mockResolvedValue(cinemaworldResponse)
        getFilmworldMovies.mockResolvedValue({
            error: new HTTPError('error', 500),
            data: [],
        })

        await expect(getMovies()).resolves.toMatchSnapshot()
    })

    test('one of the sources returns empty data', async () => {
        getCinemaworldMovies.mockResolvedValue(cinemaworldResponse)
        getFilmworldMovies.mockResolvedValue({
            error: null,
            data: [],
        })

        await expect(getMovies()).resolves.toMatchSnapshot()
    })

    test('both sources fail the request', async () => {
        getCinemaworldMovies.mockResolvedValue({
            error: new HTTPError('error', 500),
            data: [],
        })
        getFilmworldMovies.mockResolvedValue({
            error: new HTTPError('error', 500),
            data: [],
        })

        await expect(getMovies()).rejects.toThrow(HTTPError)
    })

    test('both sources return empty data sets', async () => {
        getCinemaworldMovies.mockResolvedValue({
            error: null,
            data: [],
        })
        getFilmworldMovies.mockResolvedValue({
            error: null,
            data: [],
        })

        await expect(getMovies()).rejects.toThrow(HTTPError)
    })
})
