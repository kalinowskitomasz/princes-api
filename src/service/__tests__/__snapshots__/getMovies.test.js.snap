// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getMovies correct responses from both only once source 1`] = `
Array [
  Object {
    "ID": "tt2488496",
    "Poster": "https://m.media-amazon.com/images/M/MV5BOTAzODEzNDAzMl5BMl5BanBnXkFtZTgwMDU1MTgzNzE@._V1_SX300.jpg",
    "Title": "Star Wars: Episode VII - The Force Awakens",
    "Type": "movie",
  },
  Object {
    "ID": "tt2527336",
    "Poster": "https://m.media-amazon.com/images/M/MV5BMjQ1MzcxNjg4N15BMl5BanBnXkFtZTgwNzgwMjY4MzI@._V1_SX300.jpg",
    "Title": "Star Wars: Episode VIII - The Last Jedi",
    "Type": "movie",
  },
  Object {
    "ID": "tt2527338",
    "Poster": "https://m.media-amazon.com/images/M/MV5BMDljNTQ5ODItZmQwMy00M2ExLTljOTQtZTVjNGE2NTg0NGIxXkEyXkFqcGdeQXVyODkzNTgxMDg@._V1_SX300.jpg",
    "Title": "Star Wars: Episode IX - The Rise of Skywalker",
    "Type": "movie",
  },
  Object {
    "ID": "tt3748528",
    "Poster": "https://m.media-amazon.com/images/M/MV5BMjEwMzMxODIzOV5BMl5BanBnXkFtZTgwNzg3OTAzMDI@._V1_SX300.jpg",
    "Title": "Rogue One: A Star Wars Story",
    "Type": "movie",
  },
  Object {
    "ID": "tt3778644",
    "Poster": "https://m.media-amazon.com/images/M/MV5BOTM2NTI3NTc3Nl5BMl5BanBnXkFtZTgwNzM1OTQyNTM@._V1_SX300.jpg",
    "Title": "Solo: A Star Wars Story",
    "Type": "movie",
  },
  Object {
    "ID": "tt0076759",
    "Poster": "https://m.media-amazon.com/images/M/MV5BNzVlY2MwMjktM2E4OS00Y2Y3LWE3ZjctYzhkZGM3YzA1ZWM2XkEyXkFqcGdeQXVyNzkwMjQ5NzM@._V1_SX300.jpg",
    "Title": "Star Wars: Episode IV - A New Hope",
    "Type": "movie",
  },
  Object {
    "ID": "tt0080684",
    "Poster": "https://m.media-amazon.com/images/M/MV5BYmU1NDRjNDgtMzhiMi00NjZmLTg5NGItZDNiZjU5NTU4OTE0XkEyXkFqcGdeQXVyNzkwMjQ5NzM@._V1_SX300.jpg",
    "Title": "Star Wars: Episode V - The Empire Strikes Back",
    "Type": "movie",
  },
  Object {
    "ID": "tt0086190",
    "Poster": "https://m.media-amazon.com/images/M/MV5BOWZlMjFiYzgtMTUzNC00Y2IzLTk1NTMtZmNhMTczNTk0ODk1XkEyXkFqcGdeQXVyNTAyODkwOQ@@._V1_SX300.jpg",
    "Title": "Star Wars: Episode VI - Return of the Jedi",
    "Type": "movie",
  },
  Object {
    "ID": "tt0120915",
    "Poster": "https://m.media-amazon.com/images/M/MV5BYTRhNjcwNWQtMGJmMi00NmQyLWE2YzItODVmMTdjNWI0ZDA2XkEyXkFqcGdeQXVyNTAyODkwOQ@@._V1_SX300.jpg",
    "Title": "Star Wars: Episode I - The Phantom Menace",
    "Type": "movie",
  },
  Object {
    "ID": "tt0121765",
    "Poster": "https://m.media-amazon.com/images/M/MV5BMDAzM2M0Y2UtZjRmZi00MzVlLTg4MjEtOTE3NzU5ZDVlMTU5XkEyXkFqcGdeQXVyNDUyOTg3Njg@._V1_SX300.jpg",
    "Title": "Star Wars: Episode II - Attack of the Clones",
    "Type": "movie",
  },
  Object {
    "ID": "tt0121766",
    "Poster": "https://m.media-amazon.com/images/M/MV5BNTc4MTc3NTQ5OF5BMl5BanBnXkFtZTcwOTg0NjI4NA@@._V1_SX300.jpg",
    "Title": "Star Wars: Episode III - Revenge of the Sith",
    "Type": "movie",
  },
]
`;

exports[`getMovies correct responses from both sources 1`] = `
Array [
  Object {
    "ID": "tt2488496",
    "Poster": "https://m.media-amazon.com/images/M/MV5BOTAzODEzNDAzMl5BMl5BanBnXkFtZTgwMDU1MTgzNzE@._V1_SX300.jpg",
    "Title": "Star Wars: Episode VII - The Force Awakens",
    "Type": "movie",
  },
  Object {
    "ID": "tt2527336",
    "Poster": "https://m.media-amazon.com/images/M/MV5BMjQ1MzcxNjg4N15BMl5BanBnXkFtZTgwNzgwMjY4MzI@._V1_SX300.jpg",
    "Title": "Star Wars: Episode VIII - The Last Jedi",
    "Type": "movie",
  },
  Object {
    "ID": "tt2527338",
    "Poster": "https://m.media-amazon.com/images/M/MV5BMDljNTQ5ODItZmQwMy00M2ExLTljOTQtZTVjNGE2NTg0NGIxXkEyXkFqcGdeQXVyODkzNTgxMDg@._V1_SX300.jpg",
    "Title": "Star Wars: Episode IX - The Rise of Skywalker",
    "Type": "movie",
  },
  Object {
    "ID": "tt3748528",
    "Poster": "https://m.media-amazon.com/images/M/MV5BMjEwMzMxODIzOV5BMl5BanBnXkFtZTgwNzg3OTAzMDI@._V1_SX300.jpg",
    "Title": "Rogue One: A Star Wars Story",
    "Type": "movie",
  },
  Object {
    "ID": "tt3778644",
    "Poster": "https://m.media-amazon.com/images/M/MV5BOTM2NTI3NTc3Nl5BMl5BanBnXkFtZTgwNzM1OTQyNTM@._V1_SX300.jpg",
    "Title": "Solo: A Star Wars Story",
    "Type": "movie",
  },
  Object {
    "ID": "tt0076759",
    "Poster": "https://m.media-amazon.com/images/M/MV5BNzVlY2MwMjktM2E4OS00Y2Y3LWE3ZjctYzhkZGM3YzA1ZWM2XkEyXkFqcGdeQXVyNzkwMjQ5NzM@._V1_SX300.jpg",
    "Title": "Star Wars: Episode IV - A New Hope",
    "Type": "movie",
  },
  Object {
    "ID": "tt0080684",
    "Poster": "https://m.media-amazon.com/images/M/MV5BYmU1NDRjNDgtMzhiMi00NjZmLTg5NGItZDNiZjU5NTU4OTE0XkEyXkFqcGdeQXVyNzkwMjQ5NzM@._V1_SX300.jpg",
    "Title": "Star Wars: Episode V - The Empire Strikes Back",
    "Type": "movie",
  },
  Object {
    "ID": "tt0086190",
    "Poster": "https://m.media-amazon.com/images/M/MV5BOWZlMjFiYzgtMTUzNC00Y2IzLTk1NTMtZmNhMTczNTk0ODk1XkEyXkFqcGdeQXVyNTAyODkwOQ@@._V1_SX300.jpg",
    "Title": "Star Wars: Episode VI - Return of the Jedi",
    "Type": "movie",
  },
  Object {
    "ID": "tt0120915",
    "Poster": "https://m.media-amazon.com/images/M/MV5BYTRhNjcwNWQtMGJmMi00NmQyLWE2YzItODVmMTdjNWI0ZDA2XkEyXkFqcGdeQXVyNTAyODkwOQ@@._V1_SX300.jpg",
    "Title": "Star Wars: Episode I - The Phantom Menace",
    "Type": "movie",
  },
  Object {
    "ID": "tt0121765",
    "Poster": "https://m.media-amazon.com/images/M/MV5BMDAzM2M0Y2UtZjRmZi00MzVlLTg4MjEtOTE3NzU5ZDVlMTU5XkEyXkFqcGdeQXVyNDUyOTg3Njg@._V1_SX300.jpg",
    "Title": "Star Wars: Episode II - Attack of the Clones",
    "Type": "movie",
  },
  Object {
    "ID": "tt0121766",
    "Poster": "https://m.media-amazon.com/images/M/MV5BNTc4MTc3NTQ5OF5BMl5BanBnXkFtZTcwOTg0NjI4NA@@._V1_SX300.jpg",
    "Title": "Star Wars: Episode III - Revenge of the Sith",
    "Type": "movie",
  },
]
`;

exports[`getMovies one of the sources returns empty data 1`] = `
Array [
  Object {
    "ID": "tt2488496",
    "Poster": "https://m.media-amazon.com/images/M/MV5BOTAzODEzNDAzMl5BMl5BanBnXkFtZTgwMDU1MTgzNzE@._V1_SX300.jpg",
    "Title": "Star Wars: Episode VII - The Force Awakens",
    "Type": "movie",
  },
  Object {
    "ID": "tt2527336",
    "Poster": "https://m.media-amazon.com/images/M/MV5BMjQ1MzcxNjg4N15BMl5BanBnXkFtZTgwNzgwMjY4MzI@._V1_SX300.jpg",
    "Title": "Star Wars: Episode VIII - The Last Jedi",
    "Type": "movie",
  },
  Object {
    "ID": "tt2527338",
    "Poster": "https://m.media-amazon.com/images/M/MV5BMDljNTQ5ODItZmQwMy00M2ExLTljOTQtZTVjNGE2NTg0NGIxXkEyXkFqcGdeQXVyODkzNTgxMDg@._V1_SX300.jpg",
    "Title": "Star Wars: Episode IX - The Rise of Skywalker",
    "Type": "movie",
  },
  Object {
    "ID": "tt3748528",
    "Poster": "https://m.media-amazon.com/images/M/MV5BMjEwMzMxODIzOV5BMl5BanBnXkFtZTgwNzg3OTAzMDI@._V1_SX300.jpg",
    "Title": "Rogue One: A Star Wars Story",
    "Type": "movie",
  },
  Object {
    "ID": "tt3778644",
    "Poster": "https://m.media-amazon.com/images/M/MV5BOTM2NTI3NTc3Nl5BMl5BanBnXkFtZTgwNzM1OTQyNTM@._V1_SX300.jpg",
    "Title": "Solo: A Star Wars Story",
    "Type": "movie",
  },
  Object {
    "ID": "tt0076759",
    "Poster": "https://m.media-amazon.com/images/M/MV5BNzVlY2MwMjktM2E4OS00Y2Y3LWE3ZjctYzhkZGM3YzA1ZWM2XkEyXkFqcGdeQXVyNzkwMjQ5NzM@._V1_SX300.jpg",
    "Title": "Star Wars: Episode IV - A New Hope",
    "Type": "movie",
  },
  Object {
    "ID": "tt0080684",
    "Poster": "https://m.media-amazon.com/images/M/MV5BYmU1NDRjNDgtMzhiMi00NjZmLTg5NGItZDNiZjU5NTU4OTE0XkEyXkFqcGdeQXVyNzkwMjQ5NzM@._V1_SX300.jpg",
    "Title": "Star Wars: Episode V - The Empire Strikes Back",
    "Type": "movie",
  },
  Object {
    "ID": "tt0086190",
    "Poster": "https://m.media-amazon.com/images/M/MV5BOWZlMjFiYzgtMTUzNC00Y2IzLTk1NTMtZmNhMTczNTk0ODk1XkEyXkFqcGdeQXVyNTAyODkwOQ@@._V1_SX300.jpg",
    "Title": "Star Wars: Episode VI - Return of the Jedi",
    "Type": "movie",
  },
  Object {
    "ID": "tt0120915",
    "Poster": "https://m.media-amazon.com/images/M/MV5BYTRhNjcwNWQtMGJmMi00NmQyLWE2YzItODVmMTdjNWI0ZDA2XkEyXkFqcGdeQXVyNTAyODkwOQ@@._V1_SX300.jpg",
    "Title": "Star Wars: Episode I - The Phantom Menace",
    "Type": "movie",
  },
  Object {
    "ID": "tt0121765",
    "Poster": "https://m.media-amazon.com/images/M/MV5BMDAzM2M0Y2UtZjRmZi00MzVlLTg4MjEtOTE3NzU5ZDVlMTU5XkEyXkFqcGdeQXVyNDUyOTg3Njg@._V1_SX300.jpg",
    "Title": "Star Wars: Episode II - Attack of the Clones",
    "Type": "movie",
  },
  Object {
    "ID": "tt0121766",
    "Poster": "https://m.media-amazon.com/images/M/MV5BNTc4MTc3NTQ5OF5BMl5BanBnXkFtZTcwOTg0NjI4NA@@._V1_SX300.jpg",
    "Title": "Star Wars: Episode III - Revenge of the Sith",
    "Type": "movie",
  },
]
`;
